### ISSUE TITLE
omit swarm-agent provisioning on nodes that specify `--cluster-store` and `--cluster-advertise`

### ISSUE THREAD

**Issue by dweomer** [docker/machine#3183]
... if the value of `--cluster-store` matches the value for `--swarm-discovery`.

# Problem
As @vieux [spells out in the video linked to in this article](https://blog.docker.com/2016/03/swarmweek-docker-online-meetup-35-swarm/) and @allingeek was trying to tell me a little over a week ago, swarm agents on clustered nodes are superfluous and cause the system to duplicate efforts with regards to cluster participation.

# Solution
For nodes created with `--engine-opt cluster-store ...`, `--engine-opt cluster-advertise ...`, `--swarm`, and `--swarm-discovery ...` machine should either
* refuse to create the node while emitting a helpful message as to why, or
* simply not provision the agent container while emitting an informative message as to why

# Caveats
* This mostly overlaps with #2302 but not completely. 
* This only applies the the values for the engines `--cluster-store` URL and the `--swarm-discovery` URL are equal, unless such is invalid configuration (I haven't thought that part through, Maybe a swarm maintainer could chime in here)

**Comment by nathanleclaire** [comment]
> swarm agents on clustered nodes are superfluous and cause the system to duplicate efforts with regards to cluster participation.

I don't understand the issue.  What's the duplicated effort exactly? 

libnetwork membership and swarm cluster membership are decoupled today AFAIK, I wasn't aware that you could have both without setting them up individually.  In Swarm's case it's by running the agent, in libnetwork's by setting the Docker daemon setting. @vieux @madhu maybe you can help explain?

**Comment by allingeek** [comment]
Libnetwork performs the same node registration step as swarm join. So swarm
managers will see non-libnetwork enabled engines that use swarm join and
libnetwork engines.

If you use libnetwork you don't need swarm join.

Jeff

**Comment by nathanleclaire** [comment]
How does the master become aware of the non-`join`ed members?  It happens automatically just by using the same KV store for `manage` and for `--cluster-store`?

**Comment by allingeek** [comment]
Yes. Libnetwork does the exact same thing as swarm join with respect to engine registration. Give it a shot.

— Jeff

**Comment by nathanleclaire** [comment]
Cool thanks for the speedy response @allingeek

**Comment by dweomer** [comment]
Thinking about the the values for `--cluster-store` and `--swarm-discovery` being different: makes sense to allow for such as it might be a way to
* attain greater scale (however marginal that seems at the moment)
* trash your overlay networks and start over without trashing your entire cluster
* simply maintain separation of these two KVS for the corresponding parts of your infra (stability?)

What are the other potential use-cases?

**Comment by dweomer** [comment]
Ah there is a trick, @vieux mentions it [in the video after 4:35](https://youtu.be/S5swSjF4Cn4?t=275). The paths in the KVS are different by default for swarm and libnetwork. If you pass --discovery-opt kv.path=docker/docker then the two systems should be able to share `/nodes`. I have yet to get this to work as passing a value for `--discovery-opt` via `--swarm-opt` isn't working for me.

**Comment by danhenke** [comment]
@dweomer worked for me like this --swarm-opt discovery-opt="kv.path=docker/nodes" .. Note that path is docker/nodes and not docker/docker as per your comment above.

**Comment by dweomer** [comment]
@danhenke thanks for confirming that!
